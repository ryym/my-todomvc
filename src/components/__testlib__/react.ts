import { ReactElement } from 'react';
import ReactDOM from 'react-dom';
import { act } from 'react-dom/test-utils';
import renderer from 'react-test-renderer';

// Though we used @testing-library/react first, it looks like it has a bug
// about the change event simulation (https://github.com/testing-library/react-testing-library/issues/359).
// And the official react-dom/test-utils is useful enough.

export const render = (element: ReactElement<any>) => {
  const container = document.createElement('div');
  act(() => {
    ReactDOM.render(element, container);
  });
  return container;
};

export const getByTestId = (container: HTMLElement, testId: string): HTMLElement => {
  const element = container.querySelector(`[data-testid="${testId}"]`) as HTMLElement;
  if (element == null) {
    throw new Error(`Test ID not found: ${testId}`);
  }
  return element;
};

export const renderSnapshot = (element: ReactElement<any>) => {
  return renderer.create(element).toJSON();
};
