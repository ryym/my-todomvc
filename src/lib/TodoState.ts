export namespace TodoState {
  // Todo state for filtering.
  export enum Value {
    All = 'All',
    Active = 'Active',
    Completed = 'Completed',
  }

  export enum Hash {
    All = '#/',
    Active = '#/active',
    Completed = '#/completed',
  }

  export const values = (): TodoState.Value[] => {
    return Object.values(TodoState.Value) as TodoState.Value[];
  };

  export const toHash = (state: TodoState.Value): string => {
    return TodoState.Hash[state];
  };

  export const fromHash = (hash: string): TodoState.Value => {
    const hashKeys = Object.keys(TodoState.Hash) as (keyof typeof TodoState.Hash)[];
    for (let key of hashKeys) {
      if (TodoState.Hash[key] === hash) {
        return TodoState.Value[key];
      }
    }
    return TodoState.Value.All;
  };
}
