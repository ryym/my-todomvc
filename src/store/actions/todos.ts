import { defineActions, effect } from 'redy';

let _globalTodoId = 1;

export const $todos = defineActions('todos', {
  AddNew: effect((title: string) => async dispatch => {
    if (title != null) {
      title = title.trim();
      if (title.length > 0) {
        dispatch($todos.Add(_globalTodoId++, title));
      }
    }
  }),

  Add: (id: number, title: string) => ({ id, title }),

  Remove: (id: number) => ({ id }),

  ChangeTitle: (id: number, title: string) => ({ id, title }),

  ToggleCompleted: (id: number, completed: boolean) => ({ id, completed }),

  ToggleCompletedAll: (completed: boolean) => ({ completed }),

  ClearCompleted: () => {},
});
