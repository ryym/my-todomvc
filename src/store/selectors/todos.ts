import { State } from '../state';
import { Todo } from '../../lib/Todos';
import { TodoState } from '../../lib/TodoState';
import { unreachable } from '../../lib/unreachable';

export const selectFilteredTodos = ({ todos, filters }: State): Todo[] => {
  switch (filters.todoState) {
    case TodoState.Value.All:
      return todos;

    case TodoState.Value.Active:
      return todos.filter(t => !t.completed);

    case TodoState.Value.Completed:
      return todos.filter(t => t.completed);

    default:
      return unreachable(filters.todoState);
  }
};
