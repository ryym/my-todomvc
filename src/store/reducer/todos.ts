import { defineReducer, on } from 'redy';
import { $todos } from '../actions/todos';
import { Todos } from '../../lib/Todos';

export const reduceTodos = defineReducer(Todos.init(), [
  on($todos.Add, (todos, { id, title }) => {
    const newTodo = Todos.create(id, title);
    return Todos.add(todos, newTodo);
  }),

  on($todos.Remove, (todos, { id }) => {
    return Todos.remove(todos, [id]);
  }),

  on($todos.ChangeTitle, (todos, { id, title }) => {
    const todo = Todos.find(todos, id);
    const nextTodo = { ...todo, title };
    return Todos.replace(todos, [nextTodo]);
  }),

  on($todos.ToggleCompleted, (todos, { id, completed }) => {
    const todo = Todos.find(todos, id);
    return Todos.replace(todos, [{ ...todo, completed }]);
  }),

  on($todos.ToggleCompletedAll, (todos, { completed }) => {
    const toggledTodos = todos.map(todo => ({ ...todo, completed }));
    return Todos.replace(todos, toggledTodos);
  }),

  on($todos.ClearCompleted, todos => {
    const completedIds = todos.filter(todo => todo.completed).map(t => t.id);
    return Todos.remove(todos, completedIds);
  }),
]);
