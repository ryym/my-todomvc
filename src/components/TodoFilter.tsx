import React, { useEffect } from 'react';
import classNames from 'classnames';
import { connect } from '../store/connect';
import { $app } from '../store/actions/app';
import { $todos } from '../store/actions/todos';
import { TodoState } from '../lib/TodoState';
import { unreachable } from '../lib/unreachable';

export const TodoFilter = connect(
  ({ todos, filters }) => {
    return {
      currentTodoState: filters.todoState,
      hasCompletedTodos: todos.some(t => t.completed),
    };
  },

  function TodoFilter({ dispatch, currentTodoState, hasCompletedTodos }) {
    useHashChangeEvent(hash => {
      dispatch($app.HashChanged(hash));
    });

    return (
      <>
        <ul className="filters">
          {TodoState.values().map(todoState => (
            <li key={todoState}>
              <a
                className={classNames({ selected: todoState === currentTodoState })}
                href={TodoState.toHash(todoState)}
                data-value={todoState}
              >
                {todoStateLabel(todoState)}
              </a>
            </li>
          ))}
        </ul>
        {hasCompletedTodos && (
          <button
            className="clear-completed"
            onClick={() => dispatch($todos.ClearCompleted())}
          >
            Clear completed
          </button>
        )}
      </>
    );
  }
);

const todoStateLabel = (todoState: TodoState.Value): string => {
  switch (todoState) {
    case TodoState.Value.All:
      return 'All';
    case TodoState.Value.Active:
      return 'Active';
    case TodoState.Value.Completed:
      return 'Completed';
    default:
      return unreachable(todoState);
  }
};

const useHashChangeEvent = (listener: (hash: string) => void) => {
  useEffect(() => {
    const f = () => listener(location.hash);
    window.addEventListener('hashchange', f);
    return () => window.removeEventListener('hashchange', f);
  }, []);
};
