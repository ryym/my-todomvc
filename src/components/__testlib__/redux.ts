import { Dispatch, Store, createStore } from 'redux';
import { State } from '../../store/state';

export function mockStore<S = State>(initialState: S, dispatch?: Dispatch): Store<S> {
  const noopReducer = (state: any) => state;
  let mockStore = createStore(noopReducer, initialState);
  if (dispatch) {
    mockStore = { ...mockStore, dispatch };
  }
  return mockStore;
}
