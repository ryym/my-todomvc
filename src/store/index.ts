import { createStore, applyMiddleware, compose, Store } from 'redux';
import { redyMiddleware } from 'redy';
import logger from 'redux-logger';
import { createReducer } from './reducer';
import { State } from './state';

// https://github.com/zalmoxisus/redux-devtools-extension#installation
const composeEnhancers: typeof compose =
  (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export const configureStore = (initialState: Partial<State>): Store<State> => {
  const reducer = createReducer();
  const enhancer = composeEnhancers(applyMiddleware(logger, redyMiddleware()));
  return createStore(reducer, initialState, enhancer);
};
