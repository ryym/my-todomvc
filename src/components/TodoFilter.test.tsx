import React from 'react';
import { Provider } from 'react-redux';
import { act, Simulate } from 'react-dom/test-utils';
import { spy } from 'sinon';
import { initState } from '../store/state';
import { $todos } from '../store/actions/todos';
import { TodoState } from '../lib/TodoState';
import { mockStore } from './__testlib__/redux';
import { renderSnapshot, render } from './__testlib__/react';
import { TodoFilter } from './TodoFilter';

it('matches snapshot', () => {
  const store = mockStore(initState());
  const tree = renderSnapshot(
    <Provider store={store}>
      <TodoFilter />
    </Provider>
  );
  expect(tree).toMatchSnapshot();
});

it('marks selected todo state filter', () => {
  const store = mockStore({
    ...initState(),
    filters: { todoState: TodoState.Value.Active },
  });
  const container = render(
    <Provider store={store}>
      <TodoFilter />
    </Provider>
  );

  const selectedLink = container.querySelector('a.selected') as HTMLLinkElement;
  expect(selectedLink).not.toBeNull();
  expect(selectedLink.dataset.value).toEqual(TodoState.Value.Active);
});

it('does not show clear-completed button if there are no completed todos', () => {
  const store = mockStore({
    ...initState(),
    todos: [{ id: 1, title: 'foo', completed: false }],
  });
  const container = render(
    <Provider store={store}>
      <TodoFilter />
    </Provider>
  );

  const button = container.querySelector('.clear-completed');
  expect(button).toBeNull();
});

it('clears completed todos on button click', () => {
  const dispatch = spy();
  const store = mockStore(
    {
      ...initState(),
      todos: [{ id: 1, title: 'foo', completed: true }],
    },
    dispatch
  );

  const container = render(
    <Provider store={store}>
      <TodoFilter />
    </Provider>
  );

  const button = container.querySelector('.clear-completed')!;
  act(() => {
    Simulate.click(button);
  });

  expect(dispatch.args).toEqual([[$todos.ClearCompleted()]]);
});
