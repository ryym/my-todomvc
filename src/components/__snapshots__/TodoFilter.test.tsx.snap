// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<ul
  className="filters"
>
  <li>
    <a
      className="selected"
      data-value="All"
      href="#/"
    >
      All
    </a>
  </li>
  <li>
    <a
      className=""
      data-value="Active"
      href="#/active"
    >
      Active
    </a>
  </li>
  <li>
    <a
      className=""
      data-value="Completed"
      href="#/completed"
    >
      Completed
    </a>
  </li>
</ul>
`;
